cmake/FindCython.cmake

```diff
@@ -14,41 +14,92 @@
 # Cython_FOUND - found Cython Python module
 ##
 
-execute_process(
-    COMMAND "${Python_EXECUTABLE}" -c "import Cython; print(Cython.__version__)"
-    RESULT_VARIABLE Cython_VERSION_CHECK_RESULT
-    OUTPUT_VARIABLE Cython_VERSION
-    ERROR_QUIET
-    OUTPUT_STRIP_TRAILING_WHITESPACE
-)
+macro(get_cython_version)
+  execute_process(
+      COMMAND "${Python_EXECUTABLE}" -c "import Cython; print(Cython.__version__)"
+      RESULT_VARIABLE Cython_VERSION_CHECK_RESULT
+      OUTPUT_VARIABLE Cython_VERSION
+      ERROR_QUIET
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+  )
+endmacro()
 
-set(Cython_FOUND FALSE)
+get_cython_version()
+
+if (Cython_FIND_REQUIRED)
+  set(Cython_FIND_MODE FATAL_ERROR)
+else()
+  set(Cython_FIND_MODE STATUS)
+endif()
 
 if (Cython_VERSION_CHECK_RESULT EQUAL 0)
+    set(Cython_FOUND TRUE)
     message(STATUS "Found Cython version: ${Cython_VERSION}")
     if (DEFINED Cython_FIND_VERSION)
         if(Cython_FIND_VERSION_EXACT)
             if (NOT (Cython_VERSION VERSION_EQUAL ${Cython_FIND_VERSION}))
-              message(STATUS
-                "Installing module Cython==${Cython_FIND_VERSION} in Python venv"
-              )
-              execute_process(
-                COMMAND
-                ${Python_EXECUTABLE} -m pip install Cython==${Cython_FIND_VERSION}
-              )
+              if(ENABLE_AUTO_DOWNLOAD)
```

Comment:
Why do we have this case split for Cython for either upgrading or installing?

Reply:
I implemented the full functionality expected of a FindXX module, even though we are currently only using one option. When the Cython_FIND_VERSION_EXACT variable is true (which is the case if you specified the EXACT keyword, e.g., `find_package(Cython 3.0.0 EXACT REQUIRED)`), it should install only the exact version of the Python module. Otherwise, any version of the Python module that is greater than or equal to the specified version is fine.